### PR TITLE
access cookie for UI

### PR DESCRIPTION
**Pull request by sremedios**
# Summary of changes

- [x] if the password for the check_password endpoint is correct, return {'success': true, 'access_key': '64 random bytes encoded as hex'} else, return {'success': false}
- [x] Store this access key somewhere on the server.
- [x] Store the time the key was generated
- [x] Check the headers of all incoming requests to anything except demonstration.html and /api/v1/check_password. If they don't have the right key, redirect them to demonstration.html
- [ ] When checking headers, if the key is older than 12 hours, invalidate it.
there is only one password for this.
- [x] it should be stored manually as part of the config (#28)
- [x] it should be salted and b2hashed

# How was this tested?
- [ ] TODO

### CODE REVIEW

**Comment by starfys** [comment]
i'll wait to merge config before review this since you forked off the config branch instead of master smh

**Comment by starfys** [comment]
merge conflicts dawg

File: arclightning_backend/src/router.rs [hunk 1]
```diff
@@ -28,12 +29,51 @@ fn list_files(path: PathBuf) -> Result<Vec<PathBuf>, io::Error> {
 #[derive(Debug, Clone)]
 pub struct Router {
     games_list: Arc<Mutex<HashMap<String, Game>>>,
+    static_dir: PathBuf,
+    access_key: Option<AccessKey>,
```
Comment (starfys) [inline arclightning_backend/src/router.rs]: this thing is gonna be accessed by requests so you probably want to wrap it with an arc mutex

File: arclightning_backend/src/router.rs [hunk 2]
```diff
@@ -55,14 +95,18 @@ impl hyper::service::NewService for Router {
...
             games_list: self.games_list.clone(),
+            static_dir: self.static_dir.clone(),
+            access_key: self.access_key,
         }))
     }
 }
 
 impl Router {
-    pub fn new(games_list: HashMap<String, Game>) -> Self {
+    pub fn new(games_list: HashMap<String, Game>, static_dir: PathBuf) -> Self {
```
Comment (starfys) [inline arclightning_backend/src/router.rs]: at this point we could just take `config` as a single argument

File: arclightning_backend/src/router.rs [hunk 3]
```diff
@@ -161,6 +205,102 @@ impl Router {
         Box::new(response)
     }
 
+    // Checks header of incoming request
+    // Returns destination dependent on provided key
+    fn check_header(
+        &self,
+        root: PathBuf,
+        request: Request<Body>,
+        salted_hash: Vec<u8>,
```
Comment (starfys) [inline arclightning_backend/src/router.rs]: check header won't be dealing with hashes, just the access token

**Comment by sremedios** [comment]
Steven, we may be moving to using actual cookie updation in the header instead of keeping the data as jsonerinos and passing it in the body of the response.  We will need to coordinate with front end to figure this out. c:

**Comment by sremedios** [comment]
I'm having issues with library linking and the `bin` directory.  The only way I could ensure that everything worked for now is to remove the whole `bin` directory into a top-level dir called `misc`, outside of the `arclightning_backend` folder.

File: misc/bin/set_password.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+use std::io;
+use std::path::PathBuf;
+
+fn main() -> Result<(), io::Error> {
+    // Note: This path requires that set_password be run from project root directory
+    let toml_filepath: PathBuf = ["server_config.toml"].iter().collect();
+
+    let mut config: Config = unpack_toml(&toml_filepath)?;
+
+    config.set_password("catgirls".to_string());
```
Comment (starfys) [inline misc/bin/set_password.rs]: can we read this password from stdin? preferably with this crate https://crates.io/crates/rpassword

File: misc/bin/config.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+pub fn write_toml(config: &Config, toml_filepath: &PathBuf) -> std::io::Result<()> {
+    let toml_string = toml::to_string(&config).map_err(|err| {
+        io::Error::new(
+            ErrorKind::Other,
+            format!("An error occured when serializing config toml: {}", err),
+        )
+    })?;
+    let mut file = File::create(toml_filepath)?;
```
Comment (starfys) [inline misc/bin/config.rs]: slight nit, but can we create the file before generating the toml string? Also, why is this a separate function, and not something `impl`d on Config

Comment (starfys) [reply misc/bin/config.rs]: something like `Config::write_to_path`

File: misc/bin/config.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    file.write_all(toml_string.as_bytes())
+}
+
+pub fn unpack_toml(toml_filepath: &PathBuf) -> Result<Config, io::Error> {
```
Comment (starfys) [inline misc/bin/config.rs]: Something like `Config::load` would be good here

File: arclightning_backend/src/config.rs [hunk 7]
```diff
@@ -1,16 +1,46 @@
...
     pub listen_port: u16,
     pub static_dir: PathBuf,
+    pub password: Option<String>,
     pub games: HashMap<String, Game>,
 }
 
+impl Config {
+    pub fn set_password(&mut self, password: String) -> std::io::Result<()> {
+        // The cost is set to 4 for our demo purposes to keep speed up
+        let hashed_password = hash(&password, 4).map_err(|err| {
```
Comment (starfys) [inline arclightning_backend/src/config.rs]: this should probably not be 4. let's make it `bcrypt::DEFAULT_COST`

File: arclightning_backend/src/config.rs [hunk 8]
```diff
@@ -19,13 +49,12 @@ pub struct Game {
...
     pub thumbnail_path: PathBuf,
-    #[serde(skip_serializing)]
     pub exe_path: PathBuf,
-    #[serde(skip_serializing)]
     pub exe_args: Vec<String>,
 }
 
-pub fn unpack_toml(toml_filepath: &PathBuf) -> Result<Config, io::Error> {
+
+pub fn load(toml_filepath: &PathBuf) -> Result<Config, io::Error> {
```
Comment (starfys) [inline arclightning_backend/src/config.rs]: This should be impl'd on `Config` (and called with `Config::load`)

File: arclightning_backend/src/main.rs [hunk 9]
```diff
@@ -12,9 +13,10 @@ mod password;
 mod router;
 mod tests;
 
-use config::{unpack_toml, Config, Game};
+use config::{load, Config, Game};
```
Comment (starfys) [inline arclightning_backend/src/main.rs]: remove `load` here after moving `load` to an impl

File: arclightning_backend/src/main.rs [hunk 10]
```diff
@@ -25,18 +27,15 @@ fn main() -> Result<(), io::Error> {
     let toml_filepath: PathBuf = ["server_config.toml"].iter().collect();
 
     // Unpack config
-    let config: Config = unpack_toml(&toml_filepath)?;
-    let games: HashMap<String, Game> = config.games;
-    let listen_port: u16 = config.listen_port;
-    let static_dir: PathBuf = config.static_dir;
+    let config: Config = load(&toml_filepath)?;
```
Comment (starfys) [inline arclightning_backend/src/main.rs]: `Config::load` would go here

File: arclightning_backend/src/router.rs [hunk 11]
```diff
@@ -27,14 +34,41 @@ pub fn list_files(path: PathBuf) -> Result<Vec<PathBuf>, io::Error> {
...
+#[derive(Debug, Clone)]
+struct AccessKey {
+    access_key: String,
+    access_time: u64,
+}
+
+impl AccessKey {
+    pub fn new(access_key: String, access_time: u64) -> Self {
+        AccessKey {
+            access_key: access_key.clone(),
```
Comment (starfys) [inline arclightning_backend/src/router.rs]: Why the clone here. Shouldn't passing in the string be giving ownership, making clone unneeded?

File: arclightning_backend/src/router.rs [hunk 12]
```diff
@@ -27,14 +34,41 @@ pub fn list_files(path: PathBuf) -> Result<Vec<PathBuf>, io::Error> {
...
+            access_time: access_time,
```
Comment (starfys) [inline arclightning_backend/src/router.rs]: if you're setting a field to a variable with the same name like `access_time: access_time`, you can just put `access_time`

File: arclightning_backend/src/router.rs [hunk 13]
```diff
@@ -27,14 +34,41 @@ pub fn list_files(path: PathBuf) -> Result<Vec<PathBuf>, io::Error> {
...
+        }
+    }
+    pub fn dummy() -> Self {
+        AccessKey {
+            access_key: "failure_key".to_owned(),
+            access_time: 0u64,
```
Comment (starfys) [inline arclightning_backend/src/router.rs]: is `0u64` needed here given that access time is implied to be u64?

File: misc/bin/config.rs [hunk 14]
```diff
@@ -0,0 +1,63 @@
+use bcrypt::{hash, verify};
```
Comment (starfys) [inline misc/bin/config.rs]: :thinking: is this a copy of config.rs? You should just use the original

File: misc/bin/set_password.rs [hunk 15]
```diff
@@ -0,0 +1,45 @@
+/*
```
Comment (starfys) [inline misc/bin/set_password.rs]: this should be a binary in arclightning_backend

File: arclightning_backend/src/router.rs [hunk 16]
```diff
@@ -164,14 +248,124 @@ impl Router {
...
+                )
+            }).and_then(|body| {
+                serde_json::from_slice(&body).map_err(|err| io::Error::new(ErrorKind::Other, err))
+            }).and_then(move |request_body: PasswordRequest| {
+                let password = request_body.password.clone();
+
+                // 64 random bytes encoded as hex and stored as a string
+                let mut rng = rand::thread_rng();
+                let mut session_token: String = String::new();
+                for _ in 0..64 {
```
Comment (starfys) [inline arclightning_backend/src/router.rs]: I would create a byte array and use https://docs.rs/rand/0.6.0/rand/trait.Rng.html#method.fill here, then convert the bytes to a  hex  string with the `hex` crate

File: arclightning_backend/src/router.rs [hunk 17]
```diff
@@ -164,14 +248,124 @@ impl Router {
...
+                                ErrorKind::Other,
+                                format!("Failed to acquire mutex on games list: {}", err),
+                            )
+                        })?;
+
+                        *guard = Some(AccessKey::new(
+                            session_token.clone(),
+                            SystemTime::now()
+                                .duration_since(UNIX_EPOCH)
+                                .expect("Time went backwards")
```
Comment (starfys) [inline arclightning_backend/src/router.rs]: use map_err and ? instead of expect

File: arclightning_backend/src/router.rs [hunk 18]
```diff
@@ -164,14 +248,124 @@ impl Router {
...
+                                .as_secs(),
+                        ));
+                        r#"{"success":true}"#.to_owned()
+                    } else {
+                        r#"{"success":false}"#.to_owned()
+                    };
+
+                Response::builder()
+                    .status(StatusCode::OK)
+                    .header("XSRF-TOKEN", session_token)
```
Comment (starfys) [inline arclightning_backend/src/router.rs]: TL;DR me about this

File: arclightning_backend/src/tests.rs [hunk 19]
```diff
@@ -103,3 +107,36 @@ fn test_games_serialization() {
...
+        .to_owned()
+        .into_bytes();
+
+    println!("{:?}", hashed_password);
+
+    assert!(password::check_password(password, &hashed_password));
+}
+
+#[test]
+fn test_hexdump() {
```
Comment (starfys) [inline arclightning_backend/src/tests.rs]: prob want to change this after changing the code up there maybe?

**Pull request merged**